src/api/java/com/github/minecraftschurlimods/arsmagicalegacy/api/affinity/Affinity.java

```diff
@@ -0,0 +1,201 @@
+package com.github.minecraftschurlimods.arsmagicalegacy.api.affinity;
+
+import com.github.minecraftschurlimods.arsmagicalegacy.api.ArsMagicaAPI;
+import com.github.minecraftschurlimods.arsmagicalegacy.api.util.ITranslatable;
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.sounds.SoundEvent;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.Unmodifiable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * @param color          The color for this affinity.
+ * @param minorOpposites The minor opposing affinities for this affinity.
+ * @param majorOpposites The major opposing affinities for this affinity.
+ * @param directOpposite The direct opposing affinity for this affinity.
+ * @param castSound      The sound to play when casting a spell with this affinity.
+ * @param loopSound      The sound to play when casting a continuous spell with this affinity.
+ */
+public record Affinity(int color, Set<ResourceLocation> minorOpposites, Set<ResourceLocation> majorOpposites, ResourceLocation directOpposite, Supplier<SoundEvent> castSound, Supplier<SoundEvent> loopSound) implements Comparable<Affinity>, ITranslatable {
+    public static final String AFFINITY = "affinity";
+    public static final ResourceKey<Registry<Affinity>> REGISTRY_KEY = ResourceKey.createRegistryKey(new ResourceLocation(ArsMagicaAPI.MOD_ID, AFFINITY));
+
+    public static final ResourceLocation NONE      = new ResourceLocation(ArsMagicaAPI.MOD_ID, "none");
+    public static final ResourceLocation ARCANE    = new ResourceLocation(ArsMagicaAPI.MOD_ID, "arcane");
+    public static final ResourceLocation WATER     = new ResourceLocation(ArsMagicaAPI.MOD_ID, "water");
+    public static final ResourceLocation FIRE      = new ResourceLocation(ArsMagicaAPI.MOD_ID, "fire");
+    public static final ResourceLocation EARTH     = new ResourceLocation(ArsMagicaAPI.MOD_ID, "earth");
+    public static final ResourceLocation AIR       = new ResourceLocation(ArsMagicaAPI.MOD_ID, "air");
+    public static final ResourceLocation LIGHTNING = new ResourceLocation(ArsMagicaAPI.MOD_ID, "lightning");
+    public static final ResourceLocation ICE       = new ResourceLocation(ArsMagicaAPI.MOD_ID, "ice");
+    public static final ResourceLocation NATURE    = new ResourceLocation(ArsMagicaAPI.MOD_ID, "nature");
+    public static final ResourceLocation LIFE      = new ResourceLocation(ArsMagicaAPI.MOD_ID, "life");
+    public static final ResourceLocation ENDER     = new ResourceLocation(ArsMagicaAPI.MOD_ID, "ender");
+
```

Comment:
If we're gonna do breaking API changes with this, we could just as well yeet these ResourceLocations out (except maybe for NONE). That is, unless they are used somewhere?

Reply:
They are used in creation for references to minor/major/direct opposites ... we could try and change that to Holders tho